src/Utils.vala

```diff
@@ -117,4 +118,129 @@ namespace Utils {
 
         return seat_instance;
     }
+
+    [DBus (name = "org.freedesktop.hostname1")]
+    interface HostnameInterface : Object {
+        public abstract string chassis { owned get; }
+    }
+
+    private static HostnameInterface? hostname_interface_instance;
+    private static void get_hostname_interface_instance () {
+        if (hostname_interface_instance == null) {
+            try {
+                hostname_interface_instance = Bus.get_proxy_sync (
+                    BusType.SYSTEM,
+                    "org.freedesktop.hostname1",
+                    "/org/freedesktop/hostname1"
+                );
+            } catch (GLib.Error e) {
+                warning ("%s", e.message);
+            }
+        }
+    }
+
+    private string get_chassis () {
+        get_hostname_interface_instance ();
+
+        return hostname_interface_instance.chassis;
+    }
+
+    private string? get_machine_id () {
+        string machine_id;
+        try {
+            FileUtils.get_contents ("/etc/machine-id", out machine_id);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return machine_id.strip ();
+    }
+
+    private static string? get_vendor () {
+        string vendor;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/sys_vendor", out vendor);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return vendor.strip ();
+    }
+
+    private static string? get_model () {
+        string model;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/product_name", out model);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return model.strip ();
+    }
+
+    // Based on https://git.launchpad.net/ubiquity/tree/ubiquity/misc.py?id=ae6415d224c2e76afa2274cc9f85997f38870419#n648
+    private static string? get_ubiquity_compatible_hostname () {
+        string model = get_model ();
+        string manufacturer = get_vendor ();
+
+        if (manufacturer.length == 0) {
+            return null;
+        }
+        manufacturer = manufacturer.down ();
+
+        if (manufacturer.contains ("to be filled")) {
+            // Don't bother with products in development.
+            return null;
+        }
+
+        if (manufacturer.contains ("bochs") || manufacturer.contains ("vmware")) {
+            model = "virtual machine";
+            // VirtualBox sets an appropriate system-product-name.
+        } else {
+            string key = "system-product-name";
```

Comment:
I don't get this. What is this `key` variable that's not used anywhere for?

Reply:
This was a remnant based on the call to `dmidecode`. I now read the corresponding file directly.